performance improvements

ero-qt opened the pull request.

ero-qt committed:
revert to 31 splits & performance improvements

We all hate the confusion of the 31 vs 37 split formats, one being used by runners for years and the other being used by new waves of runners. It annoys everyone greatly and users of one don't know what the other number means sometimes. We just want to go back.

Also I just made the code a bit better. Pretty sure it works for all categories (tested on normal, NGG, randomizer).

ero-qt committed:
Update Refunct.asl

ero-qt committed:
Update README.md

ero-qt committed:
Update Refunct.asl

ero-qt committed:
set all button splits to true

ero-qt committed:
Update README.md

ero-qt committed:
more descriptive variables & reintroduced brackets

ero-qt committed:
revert to original style of capitalization

DomGries commented:
Sorry I don't agree with this. This PR disables splits and removes QoL features like auto-reset after a run finished and changing to gametime by default. Also random formatting changes.

ero-qt commented:
So let me change those things? Why close it? People are complaining about lag, splits being delayed by up to 2 seconds.

BatedUrGonnaDie commented:
@DomGries by just closing this outright, you're shutting down contributions pretty hard.  If you have issues with the code, please provide a review on what you find objectionable so that it can either be changed by @just-ero or a discussion can happen over what should be done going forward.  Then at that point if an agreement can't be come too on the state of the PR close it out.  I agree with most of your points, but I can also see easy work arounds that can be accommodated for.

- Disables splits
Now that the the autosplitter has settings, it might be possible to add this in as a setting to tweak the behavior of the splits.  This way new runners can use more splits but the veterans can slim it down.
- Missing features and formatting
This should be pretty easy to add back in, assuming that everything else about the changes look good.
- Changing to gametime by default
This is definitely something that should be talked about with the rational for both sides presented and see what the community agrees upon.

Like you say, I don't think it would be good to merge this in as is for the reasons you stated, but having community patches is great and we should be trying to encourage it not shut them down.

ero-qt commented:
To add, the statement that this PR disables [some] splits is factually incorrect. It's a pity that you didn't even check my reversion of that change. And I don't think you get any say in the matter anyway, you're not part of the speedrun community. I think we all know better what we want than you do.

BatedUrGonnaDie commented:
@just-ero he is the dev of the game though, and has added quite a few things for speedrunners over the years.  Just because he isn't actively speedrunning doesn't mean he shouldn't be able to have an opinion on how things work.

ero-qt commented:
I completely disagree though. Exactly *because* he isn't actively speedrunning he shouldn't have an opinion on how things work. The additions to the game itself is a straw man argument I won't go over.

BatedUrGonnaDie commented:
@just-ero I'm not sure how speedrunning additions to the game is a straw man, but ok.  He is allowed an opinion whether you like it or not, just like I'm allowed to have one.  That's how opinions work.  How much weight they carry is an entirely different argument.

I can't speak for @DomGries , but from my perspective you keep saying "the whole community wants it this way", but there has been 0 discussion for either side on this repo prior to this.  Something like the lag/missing splits should 100% be fixed, but split behavior is very subjective on how people want it to behave.  Treat this as an open call for others to weigh in on their feelings.

From my understanding though, you want to reduce how many splits are in the run from 37 -> 31.  Is that correct?

oberien commented on Refunct/README.md:
```diff
@@ -1,22 +1,11 @@
 # Refunct Autosplitter
-Readme by @firestack and @batedurgonnadie
+(Readme by firestack, BatedUrGonnaDie, Ero)
```
You can keep the `@` to directly tag the GitHub accounts, making it easier to find them.

ero-qt replied:
Absolutely agree. This is my bad, I wasn't aware of this.

oberien replied:
Interesting, I would have imagined that to work in rendered markdown files just as it does for comments… Guess not then ¯\\\_(ツ)\_/¯

ero-qt replied:
I suppose one way to fix that is by linking to the user's page manually instead, using \[link](example.com).

BatedUrGonnaDie replied:
At this point I think it would be better to remove the by mention at the top, and move it to a general contributors section after the instructions using manual links like you said @just-ero .  That way people can add themselves when they contribute (I can add prior committers) and they can link to whatever they want to be found on.

ero-qt commented:
@BatedUrGonnaDie 
> From my understanding though, you want to reduce how many splits are in the run from 37 -> 31. Is that correct?

Absolutely not. The idea is to change the naming scheme from 37 buttons to 31 buttons with sub-buttons (7-1, 7-2, etc.). This will still keep the same amount of splits. I did want to reduce the amount of splits enabled by default to 31, since that's what it's also always been, but I suppose that change isn't necessary.

ero-qt committed:
reintroduced QoL features

Automatic change to Game Time, functional reset even when `CurrentPhase` is `Ended`.

ero-qt committed:
made instructions more concise, added users' @s

ero-qt committed:
fixed usernames

ero-qt commented:
I've made the desired changes to my repo. The README @ feature seems to not be working. I assume I made a mistake in that file that I'm missing.

ero-qt committed:
missing semicolon

BatedUrGonnaDie commented on Refunct/README.md:
```diff
@@ -1,22 +1,15 @@
-After completing a run, the autosplitter cannot reset the splits if you have beaten best times.
-Those need to be saved manually before the splitter will continue functioning.
-This is a limitation of LiveSplit.
+(Readme by @firestack, @BatedUrGonnaDie, @just-ero, @DomGries)
```
Let's just make this another top level section called "Contributors" with a list of everyone that has committed to the repo.  For now you can just add the 3 names and I can go back and collect all the rest.

ero-qt replied:
Resolved in [#2909266](https://github.com/BatedUrGonnaDie/Autosplitters/pull/28/commits/2909266fe1f0abd4a1bc793b565b761afe0c2127).

BatedUrGonnaDie commented on Refunct/Refunct.asl:
```diff
@@ -1,5 +1,4 @@
-state("Refunct-Win32-Shipping")
-{
+state("Refunct-Win32-Shipping") {
```
Let's keep the curly bracket openings on a new line just since that's what we were going with before.

This is going to apply for the rest of the file as well.

ero-qt replied:
Resolved in [#1b4690d](https://github.com/BatedUrGonnaDie/Autosplitters/pull/28/commits/1b4690db21e3f9c039a610e433379d4032b28189).

BatedUrGonnaDie commented on Refunct/Refunct.asl:
```diff
@@ -9,71 +8,69 @@ state("Refunct-Win32-Shipping")
+		if (extraButtons.ContainsKey(cluster)) {
+			for (int buttonInCluster = 1; buttonInCluster <= extraButtons[cluster]; ++buttonInCluster) {
+				settings.Add("b" + actualButtonNumber, true, "Button " + cluster + "-" + buttonInCluster, "Split on buttons:");
+				settings.SetToolTip("b" + actualButtonNumber, (buttonInCluster == 1 ? "First" : buttonInCluster == 2 ? "Second" : "Third") + " button of " + cluster);
```
Since we know this is going to be either 1, 2, or 3, I'd like to remove to double ternary for something a bit simpler and easy to read.  Maybe something like this?  I don't work with C# so I'm not sure if this is exactly valid code.
```suggestion
				settings.SetToolTip("b" + actualButtonNumber, ["First", "Second", "Third"][buttonInCluster - 1]) + " button of " + cluster);
```

ero-qt replied:
Resolved in [#1b4690d](https://github.com/BatedUrGonnaDie/Autosplitters/pull/28/commits/1b4690db21e3f9c039a610e433379d4032b28189#diff-380259e6282921ad1945b47bff8f7dac1a291d8fed1019283ab355e84083026dR33).

BatedUrGonnaDie commented on Refunct/Refunct.asl:
```diff
@@ -9,71 +8,69 @@ state("Refunct-Win32-Shipping")
 
-startup
-{
-	timer.CurrentTimingMethod = TimingMethod.GameTime;
-	vars.timerModel = new TimerModel { CurrentState = timer };
```
Is there a reason to move these out of startup and into init?

ero-qt replied:
I didn't? Those are both still in startup.

BatedUrGonnaDie replied:
I must be blind because they are clearly not in init.  Disregard this haha

BatedUrGonnaDie commented on Refunct/Refunct.asl:
```diff
@@ -9,71 +8,69 @@ state("Refunct-Win32-Shipping")
-	if (current.buttons > vars.buttons)
-	{
+split {
+	if (old.resets < current.resets && current.buttons != 0) vars.buttons = 0;
```
I think this would fit better semantically in the update block.

ero-qt replied:
Resolved in [#1b4690d](https://github.com/BatedUrGonnaDie/Autosplitters/pull/28/commits/1b4690db21e3f9c039a610e433379d4032b28189#diff-380259e6282921ad1945b47bff8f7dac1a291d8fed1019283ab355e84083026dR54).

BatedUrGonnaDie commented on Refunct/Refunct.asl:
```diff
@@ -9,71 +8,69 @@ state("Refunct-Win32-Shipping")
-			vars.timerModel.Reset();
-	}
+update {
+	if (settings.ResetEnabled && old.resets < current.resets && current.buttons == 0)
```
Is there a reason to switch to `<` vs `!=`?

ero-qt replied:
`resets` can never decrease, so `<` is all that needs to be checked.

BatedUrGonnaDie replied:
Fair enough.

BatedUrGonnaDie reviewed (changes requested):
Base changes look good, mostly style changes/questions.

BatedUrGonnaDie commented:
@just-ero it looks like you will also need to deal with the merge conflicts as well.

ero-qt committed:
moved contributors to the top

ero-qt committed:
moved bracket openings & other minor changes

`vars.buttons = 0;` to update, changed ordinal numeral choice to array instead of double ternary

ero-qt committed:
Merge branch 'master' into patch-1

BatedUrGonnaDie reviewed (approved):
Looks good to me now, thanks for the PR.
